### DISCUSSION TITLE
Can't get relevantDate working

### DISCUSSION THREAD

**Discussion by Saim-Khan1** [alexandercerutti/passkit-generator#141] [Q&A] [answered]
Hi @alexandercerutti!

Hope you are doing well. I've actually come from [scriptpapi's YouTube tutorial](https://youtu.be/rJZdPoXHtzI) on your library, was great to see you comment there as well!
So for context, I followed all the same steps shown in the video and am in essentially the same place, except with a few additional features, namely a SwiftUI DatePicker in my app, a date and time from which is stored and then sent to the backend inside the request body as a string. Here is how I am producing the string in Swift:

```
// (inside the "params" array)
"relevantDate": !(relevantDate == nil) ? StringFromDate(date: relevantDate!) : "Blank",
 "expiryDate": !(expiryDate == nil) ? StringFromDate(date: expiryDate!) : "Blank",
...

// function used to produce string from Date
func StringFromDate(date: Date) -> String {
        let dateFormatter = DateFormatter()
        dateFormatter.locale = Locale(identifier: "en_US_POSIX")
        dateFormatter.timeZone = TimeZone(abbreviation: "UTC")
        dateFormatter.dateFormat = "yyyy-MM-dd hh:mm:ss"

        return dateFormatter.string(from: date)
    }
```

And then in my backend I set the dates as follows:

```
if (request.body.relevantDate !== "Blank") {
        newPass.setRelevantDate(new Date(request.body.relevantDate));
}

if (request.body.expiryDate !== "Blank") {
        newPass.setExpirationDate(new Date(request.body.expiryDate));
}
```

But I can never get a pass to show up on the lock screen from its relevant date. Expiry date works perfectly fine, so I assume it's something to do with the time on the relevantDate, but I'm not entirely sure. I'd really appreciate any ideas you might have.

Thanks a lot!

**Comment by alexandercerutti** [comment, marked as answer]
Hey there @Saim-Khan1, thanks for using passkit-generator!

Did you perhaps open the `.pkpass` file to check if `relevantDate` has been included in `pass.json`?

Just to remember it to you, they are zip files, so you can rename them and extract their files.

Also, since I'm lacking of some information, I can make some hypothesis:

1) I was looking at the documentation. I remembered correctly that

> [Test your pass relevancy settings on a device. Simulator doesn’t show passes on the lock screen.](https://developer.apple.com/documentation/walletpasses/pass/showing_a_pass_on_the_lock_screen#:~:text=test%20your%20pass%20relevancy%20settings%20on%20a%20device.%20simulator%20doesn%E2%80%99t%20show%20passes%20on%20the%20lock%20screen.)

So, be sure you are not testing your pass on a simulator but on a real device.

2) In the legacy documentation you can find a reference about pass types. It seems that some of them do not support relevant date, but I don't know if this information is still valid. You can find the table here: https://developer.apple.com/library/archive/documentation/UserExperience/Conceptual/PassKit_PG/Creating.html#//apple_ref/doc/uid/TP40012195-CH4-SW53

So, be sure you are using a compatible one.

Let me know 👍

**Reply by Saim-Khan1** [reply]
Hey, thanks for the tip about renaming the .pkpass to a .zip to extract files, that's really useful for debugging.

1. Yep, I did see that in the documentation and have been testing on-device, thanks for pointing that out though.

2. Ahhh I didn't see that table! So it turns out that everything has been working this whole time, but Apple Wallet as you say is just really particular about which passes can and can't have dates, and whether they also need locations to show up or not. So according to that table, only boarding passes and event tickets support relevant dates by themselves (without also requiring a location).

I've just tested it now, seems to work fine for boarding passes but I can't get it working at all for event tickets? The correct date and time does come through, but it doesn't show up on the lock screen at all. Maybe that table is out of date, not entirely sure.

**Reply by alexandercerutti** [reply]
I honestly don't know why it might not work for an event ticket.
Are you sure about the date and the timezone?

Anyway, I might think to add a check in passkit-generator about adding a relevantDate.

**Reply by Saim-Khan1** [reply]
Yep I checked the pkpass files and the date and timezone are all correct, and setting the exact same date and time on a Boarding pass seems to work perfectly fine. Maybe you could give it a try if you get the chance and see if it shows up for you? Just if you have time, not a huge deal otherwise.
And yes that might be a good idea, either a check or maybe a link in the API Documentation to that table so users can check which pass types do and don't support relevant dates/locations.
Otherwise thanks a lot for your help :)

**Reply by alexandercerutti** [reply]
I'll do some checks and let you know

**Reply by alexandercerutti** [reply]
@Saim-Khan1 Hey, I've tested the situation. I can confirm that `storeCard` doesn't seems to support the relevant date.

Apple's new Apple Wallet documentation states the same:

> [A pass can be relevant on a date, at a location, or both. Set a date by adding the relevantDate key to Pass for boarding passes, event tickets, and generic passes.](https://developer.apple.com/documentation/walletpasses/pass/showing_a_pass_on_the_lock_screen#:~:text=a%20pass%20can%20be%20relevant%20on%20a%20date%2C%20at%20a%20location%2C%20or%20both.%20set%20a%20date%20by%20adding%20the%20relevantdate%20key%20to%20pass%20for%20boarding%20passes%2C%20event%20tickets%2C%20and%20generic%20passes.)


On the other side, both `boardingPass` and `eventTicket` support `relevantDate` but the last seems to get highlighted as relevant some minutes after the time, like a minute or two. I honestly don't know the reason. If I try to generate an `eventTicket` for which the relevant date is already passed by a minute, it shows immediately on my lockscreen.

So, I guess `relevantDate` shouldn't be interpreted as the exact moment of the beginning of an `eventTicket` or a `boardingPass` but an arbitrary date that should be used as a marker for the system to highlight that this pass might be relevant within a few **hours** to that user.

Also having a timezone might be more important than it seems. I generated an `eventTicket` with a relevant date and it appeared to me as an event happening 2 hours later. The reason might be due to the fact that, at the end of the timestamp, `passkit-generator` adds a `Z`, which stands for "Zulu Time" or "0 time offset" or "UTC+0". So, the time specified is relative to GMT+0 / UTC+0, and I am in UTC+1, so I can understand the offset of 1 hour, but I'm not sure why it is of two hours instead.

I have a hypothesis, but I'm not able to verify it yet: if a `passkit-generator` pass is generated based on UTC+0, for Apple it might become relevant like 2 hours before, as a hardcoded value for iOS, a minute more, a minute less.

Today I took a train and generated an Apple Wallet pass through the company application. It appeared on my lockscreen 2 to 3 hours before my train departure, maybe more.

I opened it and its `pass.json` states this:

```
"relevantDate": "2023-04-12T17:55:00+02:00"
```

The time you see is the _exact time_ the train should have left the station.
Still, I'm not sure about the `+02:00` timezone indicator. They could use it to either reduce the time before leaving or increase it.

I also have a flight planned and already downloaded its pass.

It states:

```
"relevantDate": "2023-04-21T07:25:00Z"
```

My flight leaves at 09:25 UTC+1, so 08:25 UTC+0. The pass should become relevant one hour UTF+0 before the leaving, so at 07:25 or 08:25 UTC+1.

**Reply by Saim-Khan1** [reply]
Ahh I see, thanks a lot for testing that out! I still couldn't seem to get the `eventTicket` to show up, even when the relevant date is already passed by a minute as you said, no idea why. It's not a huge deal though, at least I now know that relevancy does actually work and we have an idea of when it's supposed to show up as well. And absolutely with the `boardingPass`, that's been working perfectly and I have seen it show up, as you say, within a few hours of relevancy as the system sees fit. So that's fine, thanks a lot again for testing that out!
